beacon_node/builder_client/src/lib.rs

```diff
@@ -64,13 +74,87 @@ impl BuilderHttpClient {
             server,
             timeouts: Timeouts::new(builder_header_timeout),
             user_agent,
+            ssz_enabled: Arc::new(false.into()),
         })
     }
 
     pub fn get_user_agent(&self) -> &str {
         &self.user_agent
     }
 
+    fn fork_name_from_header(&self, headers: &HeaderMap) -> Result<Option<ForkName>, String> {
+        headers
+            .get(CONSENSUS_VERSION_HEADER)
+            .map(|fork_name| {
+                fork_name
+                    .to_str()
+                    .map_err(|e| e.to_string())
+                    .and_then(ForkName::from_str)
+            })
+            .transpose()
+    }
+
+    fn content_type_from_header(&self, headers: &HeaderMap) -> ContentType {
+        let Some(content_type) = headers.get(CONTENT_TYPE_HEADER).map(|content_type| {
+            let content_type = content_type.to_str();
+            match content_type {
+                Ok(SSZ_CONTENT_TYPE_HEADER) => ContentType::Ssz,
+                _ => ContentType::Json,
+            }
+        }) else {
+            return ContentType::Json;
+        };
+        content_type
+    }
+
+    async fn get_with_header<
+        T: DeserializeOwned + ForkVersionDecode + ForkVersionDeserialize,
+        U: IntoUrl,
+    >(
+        &self,
+        url: U,
+        timeout: Duration,
+        headers: HeaderMap,
+    ) -> Result<ForkVersionedResponse<T>, Error> {
+        let response = self
+            .get_response_with_header(url, Some(timeout), headers)
+            .await?;
+
+        let headers = response.headers().clone();
+        let response_bytes = response.bytes().await?;
+
+        let Ok(Some(fork_name)) = self.fork_name_from_header(&headers) else {
+            return serde_json::from_slice(&response_bytes).map_err(Error::InvalidJson);
+        };
+
+        let content_type = self.content_type_from_header(&headers);
+
+        match content_type {
+            ContentType::Ssz => {
+                if let Ok(mut lock) = self.ssz_enabled.write() {
+                    *lock = true
+                };
+                T::from_ssz_bytes_by_fork(&response_bytes, fork_name)
+                    .map(|data| ForkVersionedResponse {
+                        version: Some(fork_name),
+                        metadata: EmptyMetadata {},
+                        data,
+                    })
+                    .map_err(Error::InvalidSsz)
+            }
+            ContentType::Json => {
```

Comment:
I think we may as well be dynamic here and set `ssz_enabled` to `false` if we get a JSON response